### DISCUSSION TITLE
Using your own mesh in DAFoam

### DISCUSSION THREAD

**Discussion by Janfoam** [mdolab/dafoam#118] [Installation]
Thank you Friedenhe for answering my questions in the FFD thread. 

I decided to open a new discussion as while this was brought up in the FFD thread has more to do with using my own mesh and not FFD per se. This will make it easier for others who have similar problems to find the answers they need.

I tried our mesh which has 224000 cells. I noticed that in the previous DAFoam version there was an option in the runscript.py to change values in the aspect ratio, orthogonality and skewness. Please see the code snippet below. I do not find this option in the newer version of DAFoam. Can you please say where this has moved to?

With our new mesh I am getting the following error: please see the output screens below. There was a failed mesh check and right after the error:  **pyDAFoam Error: Primal solution failed for the baseline design!**. Could the check mesh be the reason for the error? I relaxed the primalMinResTol a bit: instead of 1x10^-8, I have 1x10^-6. Sometime ago, this worked after having a similar error but did not help now. 

Some points for troubleshooting "cannot find file": 
When using my own mesh the zero folder was not created as usually happens when using the mesh that comes with DAFoam. When running the simulation the following error appeared about not being able to find a file. In this case, it was necessary to create the zero folder by using the command:  cp -r 0.orig 0. Once the folder is created then the simulation will run. 

![image](https://user-images.githubusercontent.com/76228791/114003167-d05bf500-985d-11eb-9f29-a8ca3f19bbc9.png)

![image](https://user-images.githubusercontent.com/76228791/113998827-b9b39f00-9859-11eb-8bb0-d9e1fb17871a.png)

![image](https://user-images.githubusercontent.com/76228791/113998403-532e8100-9859-11eb-83e1-cac4187a9031.png)

![image](https://user-images.githubusercontent.com/76228791/113999375-478f8a00-985a-11eb-9ddb-fa22688f6758.png)

![image](https://user-images.githubusercontent.com/76228791/113999606-79a0ec00-985a-11eb-9e8f-4c39433a7cb8.png)

**Comment by friedenhe** [comment]
Your mesh has an aspect ratio of 1.6e5. This is way too high. Ideally, you want this ratio to be less than 1000. A couple of thousands are acceptable, but not 1.6e5! Check the mesh in ParaView. If you really want to bypass the mesh quality check and run it, add `"checkMeshThreshold": {"maxAspectRatio": 5e5}` to the daOptions in runScript.py

**Reply by Janfoam** [reply]
Thanks Friedenhe for your insights and patience.  The recommendation worked, but I'll need to rerun as the output files are all blank.

I am a bit confused though. I did another check on the original mesh using the native checkMesh command and the aspect ratio was 3526.xxx, but when the mesh is run in DAFoam the aspect ratio becomes more than five times that amount. Can you please explain why this is so.

The simulation ran, but for some reason the output files are all blank. What could cause this to happen?

if I decide to bypass calculating the angle of attack and just insert a figure I am sure is associated  with a particular lift based on the polar, should this consideration be applied anywhere else in DAFoam except the runscript.py?

I am still seeing that the even though the lift is fixed that it still fluctuates in the calculation. In other words, while I do fix lift to say 0.5, the iterations that follow may have it at 0.49, 0.48 etc. Can you please explain why this constraint fluctuates? 

![image](https://user-images.githubusercontent.com/76228791/114458186-2b377880-9bdf-11eb-83d9-98e1722ac6de.png)

![image](https://user-images.githubusercontent.com/76228791/114458556-9e40ef00-9bdf-11eb-9d83-60f458e60be5.png)

**Reply by friedenhe** [reply]
It is possible your mesh and FFD are not compatible. Before running the optimization load your mesh and FFD in Paraview and check them. It is possible that your FFD's size in the z-direction is much smaller than your mesh (FFD does not fully contain the mesh). Regarding the initial angle of attack, if you use the latest version of DAFoam and the latest tutorial, the solveCL will automatically run before optimization. So the initial angle of attack no longer needs to be very accurate for a specific lift. For your latest question, it is normal that the lift fluctuates during the optimization. When converged, the optimization should give you CL=0.5

**Reply by Janfoam** [reply]
Hi friedenhe, Thanks for your suggestions.

I checked the FFD setup and did not find any issues. I ran the simulation again and the same problem occurred where the following output files are blank after some iterations, namely: 
- objFuncHist.txt
- objFuncHistAvg.txt
However, what I noticed is that these output files do have information during the first iterations. The objFuncHist.txt seems to load the entire 1000 iterations even just with the first iterations appears in the opt_SLSQP.txt for the drag. It appears that once there is no longer any changes in the drag in the opt_SLSQP.txt that  objFuncHist.txt and objFuncHistAvg.txt are overwritten and become blank. Can you please explain why this would happen is this is the case? 

I have attached some files to make things a bit clearer. In the first opt_SLSQP.txt, this is a simulation that I ran using the original mesh in DAFoam for the NACA 0012 to build up experience a bit with the program before putting in a finer mesh. From this file, the drag does change there is no point at which the iteration from one to the next remains the same. 

In the second and third file, this is with our own mesh. In this case after the 10th or 15th iteration the drag does not change anymore. Could this be the reason that the objFuncHist.txt and objFuncHistAvg.txt become blank after a while? 

![image](https://user-images.githubusercontent.com/76228791/115197371-357ed880-a0f1-11eb-8180-ee6cd30f5a09.png)

![image](https://user-images.githubusercontent.com/76228791/115197888-c786e100-a0f1-11eb-8d15-e46270e0c9cd.png)
![image](https://user-images.githubusercontent.com/76228791/115198047-ee451780-a0f1-11eb-82ce-821797941c5e.png)

**Reply by friedenhe** [reply]
Don't look at objFuncHist.txt because it is the objective function history for ONE flow simulation, not optimization. That is why it got re-written every optimization iteration. Your optimization had some progress and you need to post-process the results.

**Reply by Janfoam** [reply]
Thanks for your quick response. I'll continue with some variations and see how things develop.
